### PR TITLE
[SPARK-37405][PYTHON] Inline type hints for python/pyspark/ml/feature.py

### PR DESCRIPTION
**Pull request by dchvn**
### What changes were proposed in this pull request?
Inline type hints for python/pyspark/ml/feature.py

### Why are the changes needed?
We can take advantage of static type checking within the functions by inlining the type hints.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests

### CODE REVIEW

**Comment by dchvn** [comment]
cc @HyukjinKwon , @zero323

**Comment by dchvn** [comment]
also cc @itholic

**Comment by dchvn** [comment]
cc @zero323

File: python/pyspark/ml/feature.py [hunk 1]
```diff
@@ -40,6 +53,12 @@
 from pyspark.ml.wrapper import JavaEstimator, JavaModel, JavaParams, JavaTransformer, _jvm
 from pyspark.ml.common import inherit_doc
 
+if TYPE_CHECKING:
+    from py4j.java_gateway import JavaObject  # type: ignore[import]
+
+JM = TypeVar("JM", bound=JavaTransformer)
+P = TypeVar("P", bound=Params)
```
Comment (zero323) [inline python/pyspark/ml/feature.py:60]: `P` could be safely imported from `ml._typing` if I recall correctly.

**Review by zero323 — COMMENTED** [review]
All `MLReadable` objects should have their type holes filled

https://github.com/apache/spark/blob/6149e86be874f26dbb3c2e56344a3692965e2e52/python/pyspark/ml/feature.py#L130

File: python/pyspark/ml/feature.py [hunk 2]
```diff
@@ -175,17 +194,39 @@ class Binarizer(
...
+    _input_kwargs: Dict[str, Any]
+
+    @overload
+    def __init__(
+        self,
+        *,
+        threshold: float = ...,
+        inputCol: Optional[str] = ...,
+        outputCol: Optional[str] = ...,
+    ) -> None:
```
Comment (zero323) [inline python/pyspark/ml/feature.py:206]: Obsolete `None`

File: python/pyspark/ml/feature.py [hunk 3]
```diff
@@ -175,17 +194,39 @@ class Binarizer(
...
-        inputCols=None,
-        outputCols=None,
-    ):
+        threshold: float = 0.0,
+        inputCol: Optional[str] = None,
+        outputCol: Optional[str] = None,
+        thresholds: Optional[List[float]] = None,
+        inputCols: Optional[List[str]] = None,
+        outputCols: Optional[List[str]] = None,
+    ) -> None:
```
Comment (zero323) [inline python/pyspark/ml/feature.py:229]: Ditto

File: python/pyspark/ml/feature.py [hunk 4]
```diff
@@ -271,35 +332,35 @@ class _LSHParams(HasInputCol, HasOutputCol):
         typeConverter=TypeConverters.toInt,
     )
 
-    def __init__(self, *args):
+    def __init__(self, *args: Any) -> None:
```
Comment (zero323) [inline python/pyspark/ml/feature.py:335]: Ditto. Also all similar places.

File: python/pyspark/ml/feature.py [hunk 5]
```diff
@@ -271,35 +332,35 @@ class _LSHParams(HasInputCol, HasOutputCol):
...
-    def getNumHashTables(self):
+    def getNumHashTables(self) -> int:
         """
         Gets the value of numHashTables or its default value.
         """
         return self.getOrDefault(self.numHashTables)
 
 
-class _LSH(JavaEstimator, _LSHParams, JavaMLReadable, JavaMLWritable):
+class _LSH(Generic[JM], JavaEstimator[JM], _LSHParams, JavaMLReadable, JavaMLWritable):
```
Comment (zero323) [inline python/pyspark/ml/feature.py:346]: With all annotations in `.py` files, generics should be placed at the last position.

File: python/pyspark/ml/feature.py [hunk 6]
```diff
@@ -392,24 +465,29 @@ class _BucketedRandomProjectionLSHParams:
...
     )
 
     @since("2.2.0")
-    def getBucketLength(self):
+    def getBucketLength(self) -> float:
         """
         Gets the value of bucketLength or its default value.
         """
-        return self.getOrDefault(self.bucketLength)
+        return self.getOrDefault(self.bucketLength)  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/ml/feature.py:480]: This ignore shouldn't be necessary.

File: python/pyspark/ml/feature.py [hunk 7]
```diff
@@ -926,125 +1072,135 @@ class CountVectorizer(JavaEstimator, _CountVectorizerParams, JavaMLReadable, Jav
...
+        outputCol: Optional[str] = None,
+    ) -> None:
         """
         __init__(self, \\*, minTF=1.0, minDF=1.0, maxDF=2 ** 63 - 1, vocabSize=1 << 18,\
                  binary=False, inputCol=None,outputCol=None)
         """
         super(CountVectorizer, self).__init__()
         self._java_obj = self._new_java_obj("org.apache.spark.ml.feature.CountVectorizer", self.uid)
-        kwargs = self._input_kwargs
+        kwargs = self._input_kwargs  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/ml/feature.py:1089]: `_input_kwargs` annotations should be added explicitly for classes that use them. You already have them for some of the definitions in this module.

File: python/pyspark/ml/feature.py (hunk 6 above)
Comment (dchvn) [reply python/pyspark/ml/feature.py:480]: After checking with mypy: `python/pyspark/ml/feature.py:480: error: "_BucketedRandomProjectionLSHParams" has no attribute "getOrDefault"  [attr-defined]`

**Review by zero323 — COMMENTED** [review]
Now there are:

- Unused import of `TypeVar`.
- Import from `pyspark.ml._typing` should be in type checking block.

**Comment by dchvn** [comment]
> Now there are:
> 
>     * Unused import of `TypeVar`.
> 
>     * Import from `pyspark.ml._typing` should be in type checking block.

updated, thanks

**Comment by dchvn** [comment]
> Now there are:
> 
>     * Unused import of `TypeVar`.
> 
>     * Import from `pyspark.ml._typing` should be in type checking block.

Oh, I don't know why when I push `from pyspark.ml._typing import P, JM` into TYPE_CHECKING, GitHub action has failed, hope your clarification, thanks

File: python/pyspark/ml/feature.py (hunk 6 above)
Comment (dchvn) [reply python/pyspark/ml/feature.py]: Updated

File: python/pyspark/ml/feature.py [hunk 8]
```diff
@@ -354,7 +422,13 @@ def approxNearestNeighbors(self, dataset, key, numNearestNeighbors, distCol="dis
         """
         return self._call_java("approxNearestNeighbors", dataset, key, numNearestNeighbors, distCol)
 
-    def approxSimilarityJoin(self, datasetA, datasetB, threshold, distCol="distCol"):
+    def approxSimilarityJoin(
+        self,
+        datasetA: DataFrame,
+        datasetB: DataFrame,
+        threshold: float,
+        distCol: str = "disCol",
```
Comment (zero323) [inline python/pyspark/ml/feature.py:430]: Should be `distCol` not `disCol`.

File: python/pyspark/ml/feature.py [hunk 9]
```diff
@@ -5242,104 +5997,112 @@ def setParams(
...
         """
         return self._set(outputCol=value)
 
+    @overload
+    def findSynonyms(self, word: str, num: int) -> DataFrame:
+        ...
+
+    @overload
+    def findSynonyms(self, word: Vector, num: int) -> DataFrame:
+        ...
```
Comment (zero323) [inline python/pyspark/ml/feature.py:6102]: We could skip these overloads.

File: python/pyspark/ml/feature.py [hunk 10]
```diff
@@ -5350,8 +6113,16 @@ def findSynonyms(self, word, num):
             word = _convert_to_vector(word)
         return self._call_java("findSynonyms", word, num)
 
+    @overload
+    def findSynonymsArray(self, word: str, num: int) -> List[Tuple[str, float]]:
+        ...
+
+    @overload
+    def findSynonymsArray(self, word: Vector, num: int) -> List[Tuple[str, float]]:
```
Comment (zero323) [inline python/pyspark/ml/feature.py:6121]: Ditto

File: python/pyspark/ml/feature.py [hunk 11]
```diff
@@ -6053,16 +6852,16 @@ def __init__(
...
-        selectorType="numTopFeatures",
-        percentile=0.1,
-        fpr=0.05,
-        fdr=0.05,
-        fwe=0.05,
-    ):
+        numTopFeatures: int = 50,
+        featuresCol: str = "features",
+        outputCol: Optional[str] = None,
+        labelCol: str = "label",
```
Comment (zero323) [inline python/pyspark/ml/feature.py:6858]: This is probably OK, as that's the default in constructor and on JVM side, but shouldn't be added without indicating the change.

File: python/pyspark/ml/feature.py [hunk 12]
```diff
@@ -6053,16 +6852,16 @@ def __init__(
...
+        selectorType: str = "numTopFeatures",
+        percentile: float = 0.1,
+        fpr: float = 0.05,
+        fdr: float = 0.05,
+        fwe: float = 0.05,
+    ) -> "ChiSqSelector":
         """
         setParams(self, \\*, numTopFeatures=50, featuresCol="features", outputCol=None, \
                   labelCol="labels", selectorType="numTopFeatures", percentile=0.1, fpr=0.05, \
```
Comment (zero323) [inline python/pyspark/ml/feature.py:6867]: `labelCol="labels"` -> `labelCol="label"`, see https://github.com/apache/spark/pull/35530/files#r847560912

File: python/pyspark/ml/feature.py [hunk 13]
```diff
@@ -6504,11 +7346,11 @@ def __init__(
...
         self,
         *,
-        featuresCol="features",
-        outputCol=None,
-        labelCol="labels",
-        selectionMode="numTopFeatures",
-    ):
+        featuresCol: str = "features",
+        outputCol: Optional[str] = None,
+        labelCol: str = "label",
```
Comment (zero323) [inline python/pyspark/ml/feature.py:7351]: Same as https://github.com/apache/spark/pull/35530/files#r847560912

File: python/pyspark/ml/feature.py [hunk 14]
```diff
@@ -6605,7 +7450,7 @@ def selectedFeatures(self):
     import tempfile
 
     import pyspark.ml.feature
-    from pyspark.sql import Row, SparkSession
+    from pyspark.sql import Row, SparkSession, DataFrame
```
Comment (zero323) [inline python/pyspark/ml/feature.py:7453]: This shouldn't be added.

**Comment by zero323** [comment]
Other than issues listed above, looks OK.

**Comment by itholic** [comment]
Thanks for your time to review this, @zero323 !

**Comment by itholic** [comment]
@dchvn Can you address the comments when you find some time ?? Otherwise looks good to me. 👍 

Also mind addressing https://github.com/apache/spark/pull/34324 ?? now we have only 2 PRs for completing the PySpark  type hint ticket.

Thanks for your contribution!

**Comment by dchvn** [comment]
Updated, Thanks for your time @zero323 @itholic

**Comment by dchvn** [comment]
GA has failed, so check the tests at my forked repo: https://github.com/dchvn/spark/actions/runs/2165657489
Thanks

**Comment by zero323** [comment]
Merged into master and branch-3.3.

I am going to handle https://github.com/apache/spark/pull/35530#discussion_r847442480 in a follow-up.

**Comment by dchvn** [comment]
Thanks @zero323

**Pull request closed without merging**